src/systems/collision.rs

```diff
@@ -23,3 +29,109 @@ impl<'s> System<'s> for EnforceBoundsSystem {
         }
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct CollisionEvent {
+    pub entity_a: Entity,
+    pub entity_b: Entity,
+}
+
+impl CollisionEvent {
+    pub fn new(entity_a: Entity, entity_b: Entity) -> CollisionEvent {
+        CollisionEvent { entity_a, entity_b }
+    }
+}
+
+pub struct CollisionSystem;
+
+impl<'s> System<'s> for CollisionSystem {
+    type SystemData = (
+        ReadStorage<'s, collider::Circle>,
+        WriteStorage<'s, creatures::Movement>,
+        WriteStorage<'s, Transform>,
+        Entities<'s>,
+        Write<'s, EventChannel<CollisionEvent>>,
+    );
+
+    fn run(
+        &mut self,
+        (circles, mut movements, locals, entities, mut collision_events): Self::SystemData,
+    ) {
+        for (circle_a, movement, local_a, entity_a) in
+            (&circles, &mut movements, &locals, &entities).join()
+        {
+            for (circle_b, local_b, entity_b) in (&circles, &locals, &entities).join() {
+                if entity_a == entity_b {
+                    continue;
+                }
+
+                let allowed_distance = circle_a.radius + circle_b.radius;
+                let direction = local_a.translation() - local_b.translation();
+                if direction.magnitude_squared() < allowed_distance * allowed_distance {
+                    collision_events.single_write(CollisionEvent::new(entity_a, entity_b));
+
+                    if direction.magnitude() < f32::EPSILON {
```

Comment:
Will probably later want to split off the resolution of collisions into another system. For now I think this is good!